lib/Parse/ParseDecl.cpp

```diff
@@ -2936,6 +2946,33 @@ Parser::parseDecl(ParseDeclOptions Flags,
 
   // Obvious nonsense.
   default:
+
+    if (Flags.contains(PD_HasContainerType) &&
+        IsAtStartOfLineOrPreviousHadSemi) {
+
+      // Emit diagnostics if we meet an identifier/operator where a declaration
+      // is expected, perhaps the user forgot the 'func' or 'var' keyword.
+      //
+      // Must not confuse it with trailing closure syntax, so we only
+      // recover in contexts where there can be no statements.
+
+      if ((Tok.isIdentifierOrUnderscore() &&
+           peekToken().isAny(tok::colon, tok::equal, tok::comma)) ||
+          Tok.is(tok::l_paren)) {
```

Comment:
The l-paren case seems very subtle. Maybe if it were an lparen and the *next* token were an identifier too, but otherwise I'm not sure I trust this as a recovery.

Reply:
At first I thought that that wouldn't catch nested tuples such as `var ((x, y), z) = ((1, 2), 3)`, and I even included it as a test case, but now I see that this is probably very rare and indeed not worth it.

So yeah, I'll change it according to your comment.